truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalInFrameNode.java

```diff
@@ -0,0 +1,201 @@
+package org.truffleruby.language.threadlocal;
+
+import org.truffleruby.Layouts;
+import org.truffleruby.RubyContext;
+import org.truffleruby.language.RubyBaseNode;
+import org.truffleruby.language.arguments.RubyArguments;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.Frame;
+import com.oracle.truffle.api.frame.FrameDescriptor;
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.frame.FrameSlotKind;
+import com.oracle.truffle.api.frame.MaterializedFrame;
+import com.oracle.truffle.api.object.DynamicObject;
+
+public abstract class ThreadLocalInFrameNode extends RubyBaseNode {
+
+    protected final String variableName;
+    protected final int limit;
+
+    public ThreadLocalInFrameNode(String variableName, int limit) {
+        this.variableName = variableName;
+        this.limit = limit;
+    }
+
+    public abstract ThreadLocalObject execute(MaterializedFrame frame);
+
+    @Specialization(guards = "strategy.matches(frame)", limit = "limit")
+    public ThreadLocalObject getThreadLocalCached(MaterializedFrame frame,
+            @Cached("of(getContext(), frame, variableName)") SetThreadLocalInFrameStrategy strategy) {
+        return strategy.getThreadLocal(getContext(), frame);
+    }
+
+    @Specialization(replaces = "getThreadLocalCached")
+    public ThreadLocalObject getThreadLocal(MaterializedFrame frame) {
+        return getThreadLocalSearchingDeclarations(getContext(), frame, variableName);
+    }
+
+
+    public static abstract class SetThreadLocalInFrameStrategy {
+        protected final FrameDescriptor descriptor;
+        protected final FrameSlot slot;
+        protected final Object defaultValue;
+
+        protected SetThreadLocalInFrameStrategy(FrameDescriptor fd, FrameSlot fs, Object defaultValue) {
+            this.descriptor = fd;
+            this.slot = fs;
+            this.defaultValue = defaultValue;
+        }
+
+        public boolean matches(Frame callerFrame) {
+            return callerFrame != null && callerFrame.getFrameDescriptor() == descriptor;
+        }
+
+        public boolean matchesBlock(DynamicObject block) {
+            return matches(Layouts.PROC.getDeclarationFrame(block));
+        }
+
+        public abstract ThreadLocalObject getThreadLocal(RubyContext context, MaterializedFrame callerFrame);
+
+        public ThreadLocalObject getThreadLocalBlock(RubyContext context, DynamicObject block) {
+            return getThreadLocal(context, Layouts.PROC.getDeclarationFrame(block).materialize());
+        }
+
+        public static SetThreadLocalInFrameStrategy ofBlock(RubyContext context, DynamicObject block, String varName) {
```

Comment:
This seems to never be used anymore, so I think the whole strategy abstraction can be removed.